Data driven new project wizard

clrudolphi opened the pull request:
### 🤔 What's changed?

Experimental PR to explore ways to make the new project wizard data-driven and pull dependency information from an http json file.

### ⚡️ What's your motivation? 

<!-- 
What motivated you to propose this change? Does it fix a bug? Add a new feature?
If it fixes an open issue, you can link to the issue here, e.g. "Fixes #99"
-->

### 🏷️ What kind of change is this?

<!--- Delete any options that are not relevant -->

- :book: Documentation (improvements without changing code)
- :bank: Refactoring/debt/DX (improvement to code design, tooling, etc. without changing behaviour)
- :bug: Bug fix (non-breaking change which fixes a defect)
- :zap: New feature (non-breaking change which adds new behaviour)
- :boom: Breaking change (incompatible changes to the API)

### ♻️ Anything particular you want feedback on?

<!-- 
Is there anything in this change you're unsure about, or would 
particularly like reviewers to give you feedback on?
-->

### 📋 Checklist:

<!--- 
This is to help you remember all the little things we often forget to do!

Feel free to delete any tasks that are not relevant, or add new ones.
-->

- [ ] I've changed the behaviour of the code
  - [ ] I have added/updated tests to cover my changes.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [ ] Users should know about my change
  - [ ] I have added an entry to the "[vNext]" section of the [**CHANGELOG**](../blob/main/CHANGELOG.md), linking to this pull request & included my GitHub handle to the release contributors list.

----

*This text was originally taken from the [template of the Cucumber project](https://github.com/cucumber/.github/blob/main/.github/PULL_REQUEST_TEMPLATE.md), then edited by hand. [You can modify the template here.](https://github.com/reqnroll/.github/edit/main/.github/PULL_REQUEST_TEMPLATE.md)*

gasparnagy commented on Reqnroll.VisualStudio/Resources/TestFrameworkDescriptors.json:
```diff
@@ -0,0 +1,53 @@
+        "name": "Reqnroll.NUnit",
+        "version": "2.4.1"
+      },
+      "TestFramework_Lib": {
```
I fear that there will be test frameworks that require more than one dependencies (maybe in some combinations we might need even multiple Reqnroll dependencies too), so I think I would prefer having a general list of dependencies that we just add and not have that categorization.

clrudolphi replied:
I understand your rationale. When allowing for a more flexible set of package references, the wizard code must then generate a `<PackageReference>` string for each and the number of them can't be predicted ahead of time. That means the template would have a single substitution token into which all of the package reference strings must go.

I've spent all afternoon on this and keep running in to the same problem.
Template substitution works fine when the substitution variable represents a single xml element to be injected into the csproj (or a subsection of an element's textual representation).

When there are multiple items to be injected, each has to have its own substitution variable. Any attempt to include more than one line or more than one xml element causes an error.

I've attempted every way I can think of to join multiple together, but MSBuild keeps complaining that there is invalid xml (specifically 'The element <#text> beneath element `<ItemGroup>` is unrecognized'). I've joined the package reference strings with `Environment.NewLine`, with a hard-coded '\n' and with no separation at all. All result in the same error.

Have you seen this problem before? (I'm struggling to find reliable docs and information on what is/is-not allowed)

Code-Grump replied:
It sounds like the system is rejecting XML fragments. Could you add a complete `<ItemGroup>` with all its ancestors?

clrudolphi replied:
It is rejecting multi-line substitution variable values.
I tried the suggestion of substituting the entire `<ItemGroup>` element, but ended up with the same error.

clrudolphi commented:
Here is a working example. I went with an approach that supports up to 10 PackageReferences supported. Each reference gets its own line in the csproj template and a flag indicating whether it has a value.
Admittedly, it is a bit of a hack and I can go back and try a few other things, but the other approaches I have in mind bypass the templating and directly manipulate the Project model in the DTE.

gasparnagy commented:
> Here is a working example. I went with an approach that supports up to 10 PackageReferences supported. Each reference gets its own line in the csproj template and a flag indicating whether it has a value. Admittedly, it is a bit of a hack and I can go back and try a few other things, but the other approaches I have in mind bypass the templating and directly manipulate the Project model in the DTE.

I think this might have been some other issue. I have tried with single param and it seemed to work. Please check my commit. It's just a dummy implementation, but seems to work.

clrudolphi commented:
So, let's talk about revisions to the content of the dialog box. We have the 'description' field that we could use. We could add a list of possible validation frameworks to use in a dropdown, etc.
Here is the as-is:
![NewProjAsIs](https://github.com/user-attachments/assets/91f1e3be-86e9-4e6f-be07-c4c7a0b4c5b1)

Here is a mock-up of what it could look like:
![NewProjToBe](https://github.com/user-attachments/assets/19a4b05d-c860-4d4d-a4a8-5a472601bd6d)

Do you want to add these new features to the dialog?
Where would you like to see the placement of the Test Framework description text? (Appended to the name of the test framework as shown in the mock-up, or perhaps as a line of text either below or to the right of the dropdown that would get updated when a new selection is made)
 
Should each dropdown entry include the version number, and we provide for multiple versions in the dropdown of a given item? (In case an enterprise has standardized on a version they can select that instead of getting always the latest. I mean, within reason, no need to keep more than just a handful of each item.)

clrudolphi commented:
Added implementation as discussed. I'm not a xaml expert, only dabbled in it years ago. I don't have the xaml designer workload installed on my VS installation either. So, it's likely that I've mucked things up a bit. Please review and change what you don't like.
There is one problem that I can't seem to solve - the framework description TextBlock is displaying black text on a black background on my dark-themed VS installation. I could use some help with that.
I'll continue work by adding unit tests for the classes I've added.

gasparnagy commented:
I have "fixed" the styling by replacing the TextBlock with label. I don't remember fully how the theme support works (I remember it was quite a hack), and I could not figure out how to apply it for the TextBlock.

The data binding is still has some issues, but I don't have more time right now. I will continue on that tomorrow.

gasparnagy commented:
> Getting some strange test failures in the tests of the connectors (I haven't changed any of that code, so I don't know whats going on, would you check please?)

For me they pass. What error do you get?

clrudolphi commented:
I modified the tests to adapt to the new execution model. 
I also extended the concept of tag for the TestFrameworks thoughout. 

Take a look when you have time.
